File: core/Wallet.js
import {CustomRuntimeError} from "../utils.js";

export default class Wallet {
    isOpen;
    balance;
    cards = new Set();

    constructor() {
        this.isOpen = false;
        this.balance = 0.0;
        console.log(`initialized ${this.toString()}`)
    }

    toString() {
        return `Wallet : {isOpen : ${this.isOpen}, balance : ${this.balance}, cards : ${Array.from(this.cards)}}`
    }

    open = () => {
        this.isOpen = true;
        console.log(`Wallet opened`)
    }

    close = () => {
        this.isOpen = false
        console.log(`Wallet closed`)
    }

    getBalance = () => {
        this.checkWalletIsOpen();
        return this.balance;
    }

    withdraw = (amount) => {
        this.checkWalletIsOpen();

        if (amount > this.balance) {
            throw new CustomRuntimeError(`Error: Insufficient funds`)
        }

        this.balance -= amount;
        console.log(`Withdrawal of $ ${amount} successful. Remaining balance: $ ${this.balance}`)

        return this.balance;
    }

    deposit = (amount) => {
        this.checkWalletIsOpen();

        if (amount < 0) {
            throw new CustomRuntimeError(`Error: Invalid amount ${amount}`);
        }

        this.balance += amount;
        console.log(`Deposit of $ ${amount} successful. Remaining balance: $ ${this.balance}`);

        return this.balance;
    }

    storeCard = (cardName) => {
        this.checkWalletIsOpen();

        this.cards.add(cardName);
        console.log(`Card ${cardName} stored in the wallet.`)
        return cardName;
    }

    getCard = (cardName) => {
        this.checkWalletIsOpen();

        if (this.cards.has(cardName)) {
            console.log(`Card ${cardName} retrieved from the wallet.`)
            this.cards.delete(cardName)
            return cardName;
        }
        throw new CustomRuntimeError(`cardName ${cardName} not found.`)
    }

    getCards = () => {
        this.checkWalletIsOpen();
        return this.cards;
    }

    checkWalletIsOpen= () => {
        if (!this.isOpen) {
            throw new CustomRuntimeError(`Error: Wallet is closed.`);
        }
    }
}

File: core/Wallet.py
from exception.CustomException import CustomException


class Wallet:
    def __init__(self):
        self.is_open = False
        self.balance = 0.0
        self.cards = set()

    def open(self):
        self.is_open = True

    def close(self):
        self.is_open = False

    def get_balance(self):
        self.check_action_validity()
        print(f"Balance: $ {self.balance}")
        return

    def withdraw(self, amount: float):
        self.check_action_validity()
        if amount > self.balance:
            raise CustomException("Error: Insufficient funds.")
        self.balance = self.balance - amount
        print(f"Withdrawal of $ {amount} successful. Remaining balance: $ {self.balance}")

    def store_card(self, card_name):
        self.check_action_validity()
        if card_name in self.cards:
            print(f"Card already present, having two cards of the same name is not yet implemented")
        self.cards.add(card_name)
        print(f"Card ${card_name} stored in the wallet.")
        return self.cards

    def deposit(self, amount: float):
        self.check_action_validity()
        if amount < 0:
            raise CustomException(f"Error: Invalid amount ${amount}")
        self.balance = self.balance + amount
        print(f"Deposit of $ {amount} successful. Remaining balance: ${self.balance}")

    def get_card(self, card_name):
        self.check_action_validity()
        if card_name in self.cards:
            print(f"Card {card_name} retrieved from the wallet.")
            self.cards.remove(card_name)
            return card_name
        raise CustomException(f"Error: Card {card_name} not found.")

    def get_cards(self):
        self.check_action_validity()
        return self.cards

    def get_state(self):
        return f"Wallet {{isOpen: {self.is_open}}}"

    def check_action_validity(self):
        if not self.is_open:
            raise CustomException("Error: Wallet is closed.")
